### ISSUE TITLE
A Question: Are there known implications for requiring InstalledVersions.php in scoper-autoload.php?

### ISSUE THREAD

**Issue by andrewjmead** [humbug/php-scoper#1034]
Hey Théo,

I had a quick question for you. Thanks for taking the time.

I was running into an issue with our WordPress plugin where it would sometimes fail to activate if a non-scoped WordPress plugin was already activated.

I've seen this error mentioned [before](https://github.com/humbug/php-scoper/issues/841). Here's the error message:

```
[17-May-2024 17:09:52 UTC] PHP Fatal error:  Cannot declare class Composer\InstalledVersions, because the name is already in use in /Users/andrewmead/Sites/installed-version-conflict/wp-content/plugins/independent-analytics-pro/vendor/composer/InstalledVersions.php on line 26
```

I ended up tracing it down to [these lines](https://github.com/humbug/php-scoper/blob/ccf574db379dc4f1ccc6e8c3d56a2e779de4cfa3/src/Autoload/ScoperAutoloadGenerator.php#L108-L111) in `ScoperAutoloadGenerator.php`. The generated version in my `scoper-autoload.php` file is:

```php
$loader = require_once __DIR__.'/autoload.php';
// Ensure InstalledVersions is available
$installedVersionsPath = __DIR__.'/composer/InstalledVersions.php';
if (file_exists($installedVersionsPath)) require_once $installedVersionsPath;
```

With this all in mind, the reason I'm getting the error seems pretty clear:

1. There's another project running first. It loads `InstalledVersions.php` and defines the `Composer\InstalledVersions` class.
2. My project runs afterwords. The line `require_once $installedVersionsPath;` loads my projects version of `InstalledVersions.php` and tries to redefine the `Composer\InstalledVersions` class. This triggers the error.

To test this theory, I removed the 4 lines above and the issue was resolved.

## The Question

Obviously, removing the lines isn't a solution, but would you be open to a change to this line:

```php
if (file_exists($installedVersionsPath)) require_once $installedVersionsPath;
```

Updating it to this:

```php
if (file_exists($installedVersionsPath) && !class_exists('Composer\InstalledVersions')) require_once $installedVersionsPath;
```

I won't pretend to know the broader implications for a change like this, but it does indeed address the problem of attempting to redefine `Composer\InstalledVersions`.

Very curious to hear your thoughts.

Thanks,
Andrew

**Comment by theofidry** [comment]
@andrewjmead I think the issue is that `Composer\InstalledVersions` should be scoped.

I'll have a more detailed look, note for myself:

- e2e 034: checks that the scoped `InstalledVersions` works
- e2e 035: contains several scoped libraries (it was to test the file collision issue IIRC), could check that the installed versions works as expected there too? – it should, just to add it explicitly

**Comment by andrewjmead** [comment]
Ah. At some point I was under the impression that I should not scope Composer itself, so I had `Composer` in `exclude-namespaces`.

Not sure why I thought that. That code had been in there for some time with no specific commit message to reference.

Thanks for taking the time. Just sponsored the project!

**Comment by andrewjmead** [comment]
@theofidry I realized I wasn't clear in my previous message, but removing `Composer` from `exclude-namespaces` did indeed fix the issue I was having.

**Comment by theofidry** [comment]
>Not sure why I thought that. That code had been in there for some time with no specific commit message to reference.

It was necessary at some point indeed, I can't remember exactly when it was removed, but it wasn't that long ago I think (1 year max?)

**Comment by theofidry** [comment]
Thanks a lot for the sponsorship!

**Comment by andrewjmead** [comment]
Hey @theofidry,

I was hoping to add one more question to this thread as it's related to the the topic above. **To summarize the above, you had helped me figure out that composer needs to be scoped.**

I recently found a WordPress plugin conflict between my plugin and GiveWP. The error I'm getting is:

```
[04-Dec-2024 17:11:28 UTC] PHP Fatal error:  Cannot declare class _PhpScoper8c59cd961363\Composer\InstalledVersions, because the name is already in use in /Users/andrewmead/Projects/installed-versions-scoped-plugin/scoped-plugin/vendor/composer/InstalledVersions.php on line 27
```

 I was able to debug and figure out that my copy of `InstalledVersions.php` was getting loaded twice. It was getting loaded by my plugin and it was getting loaded by GiveWP. When this happens, GiveWP doesn't attempt to load it's own `InstalledVersions.php` file.

So the error message makes perfect sense. The file is loaded twice, attempting to declare a class with a name that's already in use in the scoped namespace.

To figure out why this was the case, I started digging around and one thing that I noticed was `$classMap` in my `autoload_static.php` file:

```
    public static $classMap = array (
        'Composer\\InstalledVersions' => __DIR__ . '/..' . '/composer/InstalledVersions.php',
    );
```

What I found curious is that the key is `Composer\\InstalledVersions` but `InstalledVersions` is scoped so it would really be `IAWP\\Composer\\InstalledVersions`.

I'm wondering if this could be causing the issue. Both plugins have `$classMap` records with the same key of `Composer\\InstalledVersions`, so is it possible that both will try to load my copy of `InstalledVersions.php'?

Should the `$classMap` actually be what I have below?

```
    public static $classMap = array (
        'IAWP\\Composer\\InstalledVersions' => __DIR__ . '/..' . '/composer/InstalledVersions.php',
    );
```

The only change is the addition of `IAWP\\`. Making this change does fix the issue. When I add the prefix, the debugger shows my plugin correctly loading my copy of `InstalledVersions.php` and GiveWP correctly loading it's own copy.

Is adding the scoper prefix here a good idea? Is there a reason it's not already there?

Thanks for taking the time.
